src/utils/aadV2TokenProvider.ts

```diff
@@ -0,0 +1,278 @@
+
+        const authParams = new AuthParameters();
+        await authParams.ParseName(name);
+
+        if (!authParams.forceNewToken) {
+            const tokenEntry = AadV2TokenCache.GetToken(authParams.GetCacheKey());
+            if (tokenEntry && tokenEntry.SupportScopes(authParams.scopes)) {
+                return tokenEntry.Token;
+            }
+        }
+
+        if (authParams.appOnly) {
+            return await this.GetConfidentialClientToken(authParams);
+        }
+
+        const deviceCodeResponse: IDeviceCodeResponse = await this.GetDeviceCodeResponse(authParams);
+        const isDone = await this.promptForUserCode(deviceCodeResponse);
+        if (isDone) {
+            return await this.GetToken(deviceCodeResponse, authParams);
+        } else {
+            return "";
+        }
+    }
+
+    private async GetDeviceCodeResponse(authParams: AuthParameters) : Promise<IDeviceCodeResponse> {
+        const request = this.createUserCodeRequest(authParams.clientId, authParams.tenantId, authParams.scopes);
+        const response = await this._httpClient.send(request);
+
+        const bodyObject = JSON.parse(response.body);
+
+        if (response.statusCode !== 200) {
+            // Fail
+            this.processAuthError(bodyObject);
+        }
+
+        if (bodyObject.error) {  // really!?
+            this.processAuthError(bodyObject);
+        }
+
+        // Get userCode out of response body
+        const deviceCodeResponse: IDeviceCodeResponse = bodyObject;
+        return deviceCodeResponse;
+    }
+
+    private async GetToken(deviceCodeResponse: IDeviceCodeResponse, authParams: AuthParameters) : Promise<string> {
+        const request = this.createAcquireTokenRequest(authParams.clientId, authParams.tenantId, deviceCodeResponse.device_code);
+        const response = await this._httpClient.send(request);
+
+        const bodyObject = JSON.parse(response.body);
+
+        if (response.statusCode !== 200) {
+            this.processAuthError(bodyObject);
+        }
+        const tokenResponse: ITokenResponse = bodyObject;
+        AadV2TokenCache.SetToken(authParams.GetCacheKey(), tokenResponse.scope.split(' '), tokenResponse.access_token);
+
+        return tokenResponse.access_token;
+    }
+
+    private async GetConfidentialClientToken(authParams: AuthParameters): Promise<string> {
+        const request = this.createAcquireConfidentialClientTokenRequest(authParams.clientId, authParams.tenantId, authParams.clientSecret as string, authParams.appUri as string);
+        const response = await this._httpClient.send(request);
+
+        const bodyObject = JSON.parse(response.body);
+
+        if (response.statusCode !== 200) {
+            this.processAuthError(bodyObject);
+        }
+        const tokenResponse: ITokenResponse = bodyObject;
+        const scopes : string[] = [];
+        if (tokenResponse.scope) {
+            tokenResponse.scope.split(' ');
```

Comment:
The return value is ignored?

Reply:
This was an error.  Confidential clients don't support user consent for scopes, so there is no need specify the required scopes and track the consented scopes.  I removed the unnecessary code.